lib/sycamore/sycamore/connectors/elasticsearch/elasticsearch_writer.py

```diff
@@ -0,0 +1,140 @@
+        assert isinstance(target_params, ElasticTargetParams)
+        assert _narrow_list_of_doc_records(records), f"Found a bad record in {records}"
+        with self._client:
+
+            def insert_into_index():
+                for r in records:
+                    yield {
+                        "_index": target_params.index_name,
+                        "_id": r.doc_id,
+                        "properties": r.properties,
+                        "embeddings": r.embeddings,
+                    }
+
+            for success, info in parallel_bulk(self._client, insert_into_index()):  # generator must be consumed
+                if not success:
+                    print(f"Insert Operation Unsuccessful: {info}")
+
+    def create_target_idempotent(self, target_params: BaseDBWriter.TargetParams):
+        assert isinstance(target_params, ElasticTargetParams)
+        try:
+            self._client.indices.create(index=target_params.index_name, mappings=target_params.mappings, timeout="30s")
+        except ApiError as e:
+            if e.status_code == 400:
+                return
+            raise e
+
+    def get_existing_target_params(self, target_params: BaseDBWriter.TargetParams) -> "ElasticTargetParams":
+        assert isinstance(target_params, ElasticTargetParams)
+        try:
+            data = self._client.indices.get_mapping(index=target_params.index_name)
+            mappings = data[target_params.index_name]["mappings"]
+            if "properties" in mappings and "embeddings" in mappings["properties"]:
+                try:
+                    mappings["properties"]["embeddings"].pop("index_options")
+                except Exception as e:
+                    print(f"Failed to pop index_options from embeddings: {e}")
+        except ApiError as e:
+            raise e
+        return ElasticTargetParams(
+            index_name=target_params.index_name,
+            mappings=mappings,
+            flatten_properties=target_params.flatten_properties,
+        )
+
+
+@dataclass
+class ElasticDocumentRecord(BaseDBWriter.Record):
+    doc_id: str
+    properties: dict
+    embeddings: Optional[list[float]]
+
+    @classmethod
+    def from_doc(cls, document: Document, target_params: BaseDBWriter.TargetParams) -> "ElasticDocumentRecord":
+        assert isinstance(target_params, ElasticTargetParams)
+        doc_id = document.doc_id
+        embedding = document.embedding
+        if doc_id is None:
+            raise ValueError(f"Cannot write documents without a doc_id. Found {document}")
+        properties = {
+            "properties": document.properties,
+            "type": document.type,
+            "text_representation": document.text_representation,
+            "bbox": document.bbox.coordinates if document.bbox else None,
+            "shingles": document.shingles,
+        }
+        droperties = drop_types(properties, drop_empty_lists=True, drop_empty_dicts=True)
```

Comment:
do empty lists and objects also cause problems in elastic?

Reply:
No, just felt cleaner to do this